### PR TITLE
Bug 1605079 - Replace upload logic in Swift

### PR DESCRIPTION
**Pull request by badboy**
Again this PR goes against the feature branch only.
It replicates the changes #813 did for Kotlin for the Swift bindings.

Commits can be reviewed one-by-one, each one should compile on its own.
Locally all Swift tests pass, I'd like to run the tests once here after review.

### CODE REVIEW

File: glean-core/ios/Glean/Net/Upload.swift [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+enum PingUploadTask {
+    case upload(PingRequest)
+    case wait
+    case done
+}
+
+extension FfiPingUploadTask {
+    func toPingUploadTask() -> PingUploadTask {
+        switch UInt32(self.tag) {
```
[the remark this answers is not in the document]
Comment (badboy) [reply glean-core/ios/Glean/Net/Upload.swift:35]: Ha, it's good that you ask and this is probably something I 1) should research once more, 2) document in the code.

Here's what I know so far:

cbindgen generates this code:

```c
enum FfiPingUploadTask_Tag {
  FfiPingUploadTask_Upload,
  FfiPingUploadTask_Wait,
  FfiPingUploadTask_Done,
};
typedef uint8_t FfiPingUploadTask_Tag;
```

This first defines an `enum FfiPingUploadTask_Tag`, and then a typedef `FfiPingUploadTask_Tag` for `uint8_t`.
Why does it do that? Namespacing mostly. That enum is actually 4 byte big, the typedef obviously just 1. And it's different types (`enum FfiPingUploadTask_Tag` and `FfiPingUploadTask_Tag`).

Our tag is also just 1 byte big. That's enough to store our 3 different cases.

Now Swift _seems_ to do the right thing: Have both that enum type and that tag type (C code is internally "translated" into Swift compatible things).
Except I think you can't differentiate between them in naming anymore.
`FfiPingUploadTask_Tag(value _)` is the constructor for the typedef'd `UInt8`.
But that's not comparable to `FfiPingUploadTask_Upload`, which has type `FfiPingUploadTask_Tag`.

And if I do `FfiPingUploadTask_Tag(rawValue: UInt32(self.tag))` compilation crashes with `error: Abort trap: 6`.

So as an escape hatch I deal with UInt32 in the conversion, just until I have proper nameable Swift types.

And that seems all to work. The data gets passed along right (at least as best as I can see: it doesn't crash and sees the right numbers and data).

I will check again if there's any way to see the actual code Swift sees.

[the remark this answers is not in the document]
Comment (travis79) [reply glean-core/ios/Glean/Net/Upload.swift:35]: I guess I was wondering if it's possible to control the output of cbindgen to define the enum as `enum FfiPingUploadTask_Tag: UInt8 { ...` in order to simplify this one block of code.  It sounds like that is non-trivial but this is all very interesting to learn!

[the remark this answers is not in the document]
Comment (badboy) [reply glean-core/ios/Glean/Net/Upload.swift:35]: We generate C code. (Standard) C doesn't give you that possibility (but it's free to choose a smaller representation if it fits)

[the remark this answers is not in the document]
Comment (badboy) [reply glean-core/ios/Glean/Net/Upload.swift:96]: Documented.

**Comment by badboy** [comment]
Swift documentation generation failed, probably because of the problems we already fixed in master.

I intend to merge this, then merge back master into that branch (to address all merge conflicts)

**Pull request merged**